README.md
# clj-tools

FIXME: description

## How to run

```bash
lein run    # run the default program

lein run -m clj-tools.core      # run the specific program

lein run -m clj-tools.ping      # run the specific program
```


## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar clj-tools-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-03-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-03-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clj-ping/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clj-ping/compare/0.1.0...0.1.1


test/clj_ping/core_test.clj
(ns clj-ping.core-test
  (:require [clojure.test :refer :all]
            [clj-ping.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/clj_tools/core_test.clj
(ns clj-tools.core-test
  #_{:clj-kondo/ignore [:refer-all]}
  (:require [clojure.test :refer :all]
            [clj-tools.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_tools/core.clj
(ns clj-tools.core
  (:gen-class)
  (:require [clojure.java.shell :as sh]))

;; Refer to Javadoc(http://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html#isReachable(int)
;; (defn java-ping
;;   "Time an .isReachable ping to a given domain"
;;   [domain timeoutSeconds]
;;   (let [addr (java.net.InetAddress/getByName domain)
;;         start (. System (nanoTime))
;;         result (.isReachable addr (* 1000 timeoutSeconds))
;;         total (/ (double (- (. System (nanoTime)) start)) 1000000.0)]
;;     {:host domain
;;      :result result
;;      :time total}))


(defn getNanoTime []
  (. System (nanoTime)))

(defn funcElapsedMs
  "exec the f(args) and return as a map, plus key :elapsedMs"
  [f & args]
  (let [start (getNanoTime)
        rtn (apply f args)  ;; TODO, like (f args)
        end (getNanoTime)
        elapsedMs (/ (double (- end start)) 1000000.0)]
    (if (= (type {}) (type rtn))
      (assoc rtn :elapsedMs elapsedMs)
      {:rtn rtn :elapsedMs elapsedMs})))

(defn shell "exec bash shell for cmd" [cmd]
  (sh/sh "bash" "-c" cmd))

(defn nslookup "get the IP of the given host" [host]
  (.getHostAddress (java.net.InetAddress/getByName host)))


(defn shell-ping
  "using bash to call ping for a given domain"
  [host timeoutSeconds]
  (let [ip (nslookup host)
        cmd (str "ping -c 1 -W " timeoutSeconds " " ip)
        rtn (funcElapsedMs shell cmd)]
    (merge {:host host
            :timeoutSeconds timeoutSeconds
            :ip ip
            :cmd cmd
            :status (if (= 0 (:exit rtn)) "pass" "failed")}
           rtn)))




(defn showUsage []
  (println "Usage: lein run www.baidu.com 10   # ping www.abc.com with timeout 10s\n"))


(defn -main "main function" [& args]
  (let [withoutParms (if (nil? (first args)) true false)
        host (if-let [url0 (first args)] url0 "www.baidu.com")
        timeoutSeconds (if-let [timeout_sec (first (rest args))] (Integer/parseInt timeout_sec) 10)
        rtn (shell-ping host timeoutSeconds)
        ;; rtn (shell "ping -c 1 -W 10 182.61.200.7")
        ;; rtn (java-ping host timeoutSeconds)
        ]
    (println "Ping:" (:host rtn) "| Status:" (:status rtn) "| Cmd:" (:cmd rtn) "| Returns:" (:exit rtn) "| Elapsed(ms):" (:elapsedMs rtn) "| Err:" (:err rtn) "\n")
    ;; (when withoutParms (println "DEBUG:" rtn))
    (when withoutParms (showUsage))
    (System/exit 0) ;; work around, "bash -c" hangs, "refer to https://clojuredocs.org/clojure.java.shell/sh#notes
    ))


src/clj_tools/ping.clj
(ns clj-tools.ping
  (:gen-class)
  (:require [clojure.java.shell :as sh]))

;; Refer to Javadoc(http://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html#isReachable(int)
;; (defn java-ping
;;   "Time an .isReachable ping to a given domain"
;;   [domain timeoutSeconds]
;;   (let [addr (java.net.InetAddress/getByName domain)
;;         start (. System (nanoTime))
;;         result (.isReachable addr (* 1000 timeoutSeconds))
;;         total (/ (double (- (. System (nanoTime)) start)) 1000000.0)]
;;     {:host domain
;;      :result result
;;      :time total}))


(defn getNanoTime []
  (. System (nanoTime)))

(defn funcElapsedMs
  "exec the f(args) and return as a map, plus key :elapsedMs"
  [f & args]
  (let [start (getNanoTime)
        rtn (apply f args)  ;; TODO, like (f args)
        end (getNanoTime)
        elapsedMs (/ (double (- end start)) 1000000.0)]
    (if (= (type {}) (type rtn))
      (assoc rtn :elapsedMs elapsedMs)
      {:rtn rtn :elapsedMs elapsedMs})))

(defn shell "exec bash shell for cmd" [cmd]
  (sh/sh "bash" "-c" cmd))

(defn nslookup "get the IP of the given host" [host]
  (.getHostAddress (java.net.InetAddress/getByName host)))


(defn shell-ping
  "using bash to call ping for a given domain"
  [host timeoutSeconds]
  (let [ip (nslookup host)
        cmd (str "ping -c 1 -W " timeoutSeconds " " ip)
        rtn (funcElapsedMs shell cmd)]
    (merge {:host host
            :timeoutSeconds timeoutSeconds
            :ip ip
            :cmd cmd
            :status (if (= 0 (:exit rtn)) "pass" "failed")}
           rtn)))




(defn showUsage []
  (println "Usage: lein run www.baidu.com 10   # ping www.abc.com with timeout 10s\n"))


(defn -main "main function" [& args]
  (let [withoutParms (if (nil? (first args)) true false)
        host (if-let [url0 (first args)] url0 "www.baidu.com")
        timeoutSeconds (if-let [timeout_sec (first (rest args))] (Integer/parseInt timeout_sec) 10)
        rtn (shell-ping host timeoutSeconds)
        ;; rtn (shell "ping -c 1 -W 10 182.61.200.7")
        ;; rtn (java-ping host timeoutSeconds)
        ]
    (println "Ping:" (:host rtn) "| Status:" (:status rtn) "| Cmd:" (:cmd rtn) "| Returns:" (:exit rtn) "| Elapsed(ms):" (:elapsedMs rtn) "| Err:" (:err rtn) "\n")
    ;; (when withoutParms (println "DEBUG:" rtn))
    (when withoutParms (showUsage))
    (System/exit 0) ;; work around, "bash -c" hangs, "refer to https://clojuredocs.org/clojure.java.shell/sh#notes
    ))
